File: include/libqelapsedtimer.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qelapsedtimer_enums = enums;

/// https://doc.qt.io/qt-6/qelapsedtimer.html
pub const qelapsedtimer = struct {
    /// New constructs a new QElapsedTimer object.
    ///
    /// ``` other: QtC.QElapsedTimer ```
    pub fn New(other: ?*anyopaque) QtC.QElapsedTimer {
        return qtc.QElapsedTimer_new(@ptrCast(other));
    }

    /// New2 constructs a new QElapsedTimer object and invalidates the source QElapsedTimer object.
    ///
    /// ``` other: QtC.QElapsedTimer ```
    pub fn New2(other: ?*anyopaque) QtC.QElapsedTimer {
        return qtc.QElapsedTimer_new2(@ptrCast(other));
    }

    /// New3 constructs a new QElapsedTimer object.
    ///
    ///
    pub fn New3() QtC.QElapsedTimer {
        return qtc.QElapsedTimer_new3();
    }

    /// CopyAssign shallow copies `other` into `self`.
    ///
    /// ``` self: QtC.QElapsedTimer, other: QtC.QElapsedTimer ```
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QElapsedTimer_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    ///
    /// ``` self: QtC.QElapsedTimer, other: QtC.QElapsedTimer ```
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QElapsedTimer_MoveAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#clockType)
    ///
    ///
    ///
    /// Returns: ``` qelapsedtimer_enums.ClockType ```
    pub fn ClockType() i64 {
        return qtc.QElapsedTimer_ClockType();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#isMonotonic)
    ///
    ///
    pub fn IsMonotonic() bool {
        return qtc.QElapsedTimer_IsMonotonic();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#start)
    ///
    /// ``` self: QtC.QElapsedTimer ```
    pub fn Start(self: ?*anyopaque) void {
        qtc.QElapsedTimer_Start(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#restart)
    ///
    /// ``` self: QtC.QElapsedTimer ```
    pub fn Restart(self: ?*anyopaque) i64 {
        return qtc.QElapsedTimer_Restart(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#invalidate)
    ///
    /// ``` self: QtC.QElapsedTimer ```
    pub fn Invalidate(self: ?*anyopaque) void {
        qtc.QElapsedTimer_Invalidate(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#isValid)
    ///
    /// ``` self: QtC.QElapsedTimer ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.QElapsedTimer_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#nsecsElapsed)
    ///
    /// ``` self: QtC.QElapsedTimer ```
    pub fn NsecsElapsed(self: ?*anyopaque) i64 {
        return qtc.QElapsedTimer_NsecsElapsed(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#elapsed)
    ///
    /// ``` self: QtC.QElapsedTimer ```
    pub fn Elapsed(self: ?*anyopaque) i64 {
        return qtc.QElapsedTimer_Elapsed(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#hasExpired)
    ///
    /// ``` self: QtC.QElapsedTimer, timeout: i64 ```
    pub fn HasExpired(self: ?*anyopaque, timeout: i64) bool {
        return qtc.QElapsedTimer_HasExpired(@ptrCast(self), @intCast(timeout));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#msecsSinceReference)
    ///
    /// ``` self: QtC.QElapsedTimer ```
    pub fn MsecsSinceReference(self: ?*anyopaque) i64 {
        return qtc.QElapsedTimer_MsecsSinceReference(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#msecsTo)
    ///
    /// ``` self: QtC.QElapsedTimer, other: QtC.QElapsedTimer ```
    pub fn MsecsTo(self: ?*anyopaque, other: ?*anyopaque) i64 {
        return qtc.QElapsedTimer_MsecsTo(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#secsTo)
    ///
    /// ``` self: QtC.QElapsedTimer, other: QtC.QElapsedTimer ```
    pub fn SecsTo(self: ?*anyopaque, other: ?*anyopaque) i64 {
        return qtc.QElapsedTimer_SecsTo(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qelapsedtimer.html#dtor.QElapsedTimer)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QElapsedTimer ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QElapsedTimer_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qelapsedtimer.html#types
pub const enums = struct {
    pub const ClockType = enum {
        pub const SystemTime: i32 = 0;
        pub const MonotonicClock: i32 = 1;
        pub const MachAbsoluteTime: i32 = 2;
        pub const PerformanceCounter: i32 = 3;
    };
};

File: src/libqelapsedtimer.cpp
#include <QElapsedTimer>
#include <qelapsedtimer.h>
#include "libqelapsedtimer.h"
#include "libqelapsedtimer.hxx"

QElapsedTimer* QElapsedTimer_new(QElapsedTimer* other) {
    return new QElapsedTimer(*other);
}

QElapsedTimer* QElapsedTimer_new2(QElapsedTimer* other) {
    return new QElapsedTimer(std::move(*other));
}

QElapsedTimer* QElapsedTimer_new3() {
    return new QElapsedTimer();
}

void QElapsedTimer_CopyAssign(QElapsedTimer* self, QElapsedTimer* other) {
    *self = *other;
}

void QElapsedTimer_MoveAssign(QElapsedTimer* self, QElapsedTimer* other) {
    *self = std::move(*other);
}

int QElapsedTimer_ClockType() {
    return static_cast<int>(QElapsedTimer::clockType());
}

bool QElapsedTimer_IsMonotonic() {
    return QElapsedTimer::isMonotonic();
}

void QElapsedTimer_Start(QElapsedTimer* self) {
    self->start();
}

long long QElapsedTimer_Restart(QElapsedTimer* self) {
    return static_cast<long long>(self->restart());
}

void QElapsedTimer_Invalidate(QElapsedTimer* self) {
    self->invalidate();
}

bool QElapsedTimer_IsValid(const QElapsedTimer* self) {
    return self->isValid();
}

long long QElapsedTimer_NsecsElapsed(const QElapsedTimer* self) {
    return static_cast<long long>(self->nsecsElapsed());
}

long long QElapsedTimer_Elapsed(const QElapsedTimer* self) {
    return static_cast<long long>(self->elapsed());
}

bool QElapsedTimer_HasExpired(const QElapsedTimer* self, long long timeout) {
    return self->hasExpired(static_cast<qint64>(timeout));
}

long long QElapsedTimer_MsecsSinceReference(const QElapsedTimer* self) {
    return static_cast<long long>(self->msecsSinceReference());
}

long long QElapsedTimer_MsecsTo(const QElapsedTimer* self, QElapsedTimer* other) {
    return static_cast<long long>(self->msecsTo(*other));
}

long long QElapsedTimer_SecsTo(const QElapsedTimer* self, QElapsedTimer* other) {
    return static_cast<long long>(self->secsTo(*other));
}

void QElapsedTimer_Delete(QElapsedTimer* self) {
    delete self;
}
